docs/proposals/override-values-inside-structured-data-files/README.md

```diff
@@ -0,0 +1,227 @@
+know that this has succeeded?
+-->
+
++ Allow users to override specific fields within JSON or YAML structures in ConfigMaps or similar resources.
+
++ Enable operations, `replace`, `add`, and `remove` on nested fields within plaintext objects.
+
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+
+
+
+
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an administrator, I want to update specific fields within JSON or YAML files in ConfigMaps to customize application behavior without replacing the entire configuration, ensuring that my changes are minimal and targeted.
+
+#### Story 2
+
+As a developer, I want to override specific environment-related configurations stored in YAML within ConfigMaps, depending on the target deployment cluster, ensuring that my application behaves consistently across environments.
+
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? 
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
++ Introducing PlaintextObjectOverrider increases the complexity of OverridePolicy. To mitigate this, thorough documentation and examples will be provided.
+
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API Change
+
+```go
+type Overriders struct {
+	...
+	// PlaintextObjectOverrider represents the rules dedicated to handling yaml or json object overrides
+	// +optional
+	PlaintextObjectOverrider []PlaintextObjectOverrider `json:"plaintextObjectOverrider,omitempty"`
+}
+
+type PlaintextObjectOverrider struct {
```

Comment:
The `ConfigMap` often stores some configuration files of the application, such as xml, application.properties, toml, etc. commonly used in Java applications.

So in your proposal, would it be better to use `YAMLPlaintextOverrider` instead of `PlaintextObjectOverrider`? 

This way, we can later expand to other `Overrider` like `XMLPlaintextOverrider`, `TOMLPlaintextOverrider`, etc.

Reply:
If there are all these types overrider configs such as YAML, JSON, TOML，XML and etc and all of them are modified by json-merge or json-patch, maybe the API should be :
```
type Overriders struct {
	...
	// YAMLPlaintextOverrider represents the rules dedicated to handling yaml  object overrides
	// +optional
	YAMLPlaintextOverrider []PlaintextObjectOverrider `json:"yamlPlaintextOverrider,omitempty"`

	// JSONPlaintextOverrider represents the rules dedicated to handling json object overrides
	// +optional
	JSONPlaintextOverrider []PlaintextObjectOverrider `json:"jsonPlaintextOverrider,omitempty"`
}

type PlaintextObjectOverrider struct {
        // Path indicates the path of target field
        Path string `json:"path"`
        // JSONPatch represents json patch rules defined with plaintext overriders.
        JSONPatch []JSONPatch `json:"jsonPatch"`
        // MergePatch represents merge patch string.
        MergePatch string `json:"mergePatch"`
}

type JSONPatch  struct {
	// Path indicates the path of target field
	Path string `json:"path"`
	// From indicates the path of original field when operator is move and copy
        From string `json:"from"`
	// Operator indicates the operation on target field.
	// Available operators are: add, replace and remove.
	// +kubebuilder:validation:Enum=add;remove;replace
	Operator JSONPatchOperator `json:"operator"`
	// Value to be applied to target field.
	// Must be empty when operator is Remove.
	// +optional
	Value apiextensionsv1.JSON `json:"value,omitempty"`
}

// PatchOperator is the set of operators that can be used in an overrider.
type JSONPatchOperator string

// These are valid patch operators.
const (
	PatchOpAdd     JSONPatchOperator = "add"
	PatchOpRemove  JSONPatchOperator = "remove"
	PatchOpReplace  JSONPatchOperator = "replace"
	PatchOpMove  JSONPatchOperator = "move"
	PatchOpTest  JSONPatchOperator = "test"
	PatchOpCopy  JSONPatchOperator = "copy"
)

```

Reply:
+1

Reply:
Is it possible to introduce an enumerated value to indicate different data formats? In the future, only the enumerated value needs to be extended.

Reply:
>. Is it possible to introduce an enumerated value to indicate different data formats?

It might not work too well. If we introduce a field to represent the data format, that would mean different data formats would use `JSON Patch` to implement overrides. For data formats like XML, they have their own patching methods.
https://www.rfc-editor.org/rfc/rfc5261#appendix-A.6

Reply:
So if we want XML to use its own xml-patch instead of json-patch, that wouldn't scale very well. - Though that's probably a long shot. 😂

Reply:
I will write two kinds of api in my proposal.

Two solutions give different implementation.

(1) each data format has different different struct
json -> json * op -> json ->  JSONPlaintextOverrider []JSONPlaintextOverrider
yaml -> yaml * op -> yaml -> YAMLPlaintextOverrider []YAMLPlaintextOverrider
xml -> xml * op -> xml -> XMLPlaintextOverrider []XMLPlaintextOverrider
...

For example json has 5 json operations, yaml has 3 yaml operations, and xml has 4 xml operations, ...
It should be 5+3+4+.. operations in total, which is a huge number.

So, each data format has different different struct would be a good solution if we choose to use native operation for each data type.

(2)  enum
json -> json * op -> json -> enum
yaml -> json * op -> yaml -> enum
xml -> json * op -> xml -> enum
...

Because we just need the number of operations of json, so we can use enum to contain all.


For me, the operations should be tailored to each specific data format. If I’m working with XML, I believe the operations should adhere to the XML format itself, rather than converting it to JSON and applying JSON-based operations.

Reply:
I modified the proposal, and gives three api designs.

Reply:
@sophiefeifeifeiya Now yaml，toml，properties doesn't has own patch proposal. So we usually convert these data to JSON for [JSON-PATCH](https://datatracker.ietf.org/doc/html/rfc6902) in code.
If some structure data (such as XML) has their own patch rules, we can extend the API by defining their own overrider in the future.

Reply:
Okay, given that we just support data formats with JSON patch, we can continue with "Each data format has different name with same struct" api design plan https://github.com/karmada-io/karmada/pull/5449#discussion_r1740129736, as enum is hard to extend in the future. What do you guys think about this? @XiShanYongYe-Chang @chaunceyjiang 

If this one is good, then I will modify the proposal details based on this and try to implement it.

Reply:
I think we can continue to push forward as we thought.

I understand that JSON and YAML data should be the main concern of this proposal. From the perspective of solution design, it is best to reserve design for other types of data. You can explain the design idea in the current proposal.